src/components/DataAccessLayer.ts

```diff
@@ -27,10 +28,84 @@ export const getStaticRawData = (): DataMCQuestion[] => {
 };
 
 export const getConvertedStaticData = (): MCQuestion[] => {
-  return UtilConversion.convertQuestions(questions);
+  // Validate questions
+  const allDataQs: DataMCQuestion[] = getStaticRawData();
+  return UtilConversion.convertQuestions(validateMCQuestions(allDataQs));
 };
 
 export const getAllQuestionsFromApi = async (): Promise<MCQuestion[]> => {
   const allQuizzes = await NetworkCalls.getQuiz();
+
   return UtilConversion.convertQuestions(allQuizzes);
 };
+
+// helpers functions with typeguards and console warns
+// IMPORTANT: This function is also used to filter out invalid tags.
+function validateMCQuestions(allDataQs: DataMCQuestion[]): DataMCQuestion[] {
+  NetworkGuard.isMCQuestionArray(allDataQs)
+    ? console.info(
+        "%cAll questions are valid",
+        "color: #00FF00",
+        "\nTotal Questions Validated:",
+        allDataQs.length,
+      )
+    : console.warn(
+        "%cWARNING: Some questions contains invalid structure",
+        "color: #FF0000",
+      );
+
+  let invalidTags = 0;
+  let noTags = 0;
+  let invalidQs = 0;
+  let totalTags = 0;
+  const allQsLength = allDataQs.length;
+
+  for (const dataQs of allDataQs as DataMCQuestion[]) {
+    // Check question structure
+    if (!NetworkGuard.isMCQuestion(dataQs)) {
+      invalidQs++;
+    }
+
+    let { tags } = dataQs;
+
+    if (!tags || (Array.isArray(tags) && !tags.length)) {
+      noTags++;
+      continue;
+    }
+
+    totalTags += tags.length;
+
+    if (!NetworkGuard.isAllTags(tags)) {
+      const validTags = tags.filter((tag) => NetworkGuard.isTag(tag));
+      invalidTags += tags.length - validTags.length;
+      tags = validTags; // remove all invalid tags
+    }
+  }
+
+  logQsWarnings(invalidQs, allQsLength, invalidTags, totalTags, noTags);
+
+  return allDataQs;
+}
+
+function logQsWarnings(
```

Comment:
You can probably break this down more either from the top logic, but if you wanted to do it from here, then you can do something like this?

```ts

function logWarning(condition: number, message: string, ...values: any[]): void {
  if (condition) {
    console.warn(message, "color: #FF0000", ...values);
  }
}

function logQsWarnings(
  invalidQs: number,
  allQsLength: number,
  invalidTags: number,
  totalTags: number,
  noTags: number,
): void {
  logWarning(invalidQs, `Invalid Questions Received: %c${invalidQs} out of ${allQsLength}`);
  logWarning(invalidTags, `Filtering out invalid tags: %c${invalidTags} out of ${totalTags}`);
  logWarning(noTags, `Questions with no tags: %c${noTags}`);
}

```

Reply:
fixed.